### PR TITLE
GUVNOR-3487: Multi-select combo box in guided decision table only has a height of a single row

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/GUVNOR-3487

This PR also updates the widget used for ```ListBox``` to ```org.gwtbootstrap3.client.ui.ListBox```

### CODE REVIEW

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactory.java [hunk 1]
```diff
@@ -65,17 +67,23 @@
...
+    public static <T, W extends ListBox, E extends MultiValueDOMElement<T, W>, F extends MultiValueSingletonDOMElementFactory<T, W, E>> Callback<E> makeOnCreationCallback(final F factory,
+                                                                                                                                                                           final GridCell<T> cell,
+                                                                                                                                                                           final Map<String, String> enumLookups) {
+        return (e) -> {
             final W widget = e.getWidget();
-            for ( Map.Entry<String, String> lookup : enumLookups.entrySet() ) {
-                widget.addItem( lookup.getValue(),
-                                lookup.getKey() );
+
+            if (widget.isMultipleSelect()) {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactory.java]: Here's a change... if the ```ListBox``` is "multiple select" then show the lessor of the number of values or ```10```

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/datepicker/DatePickerDOMElement.java [hunk 2]
```diff
@@ -17,67 +17,21 @@
...
 import org.uberfire.ext.wires.core.grids.client.widget.grid.GridWidget;
 import org.uberfire.ext.wires.core.grids.client.widget.layer.GridLayer;
 
 /**
- * A DOMElement for TextBoxes.
+ * A DOMElement for DatePicker.
  */
 public class DatePickerDOMElement extends SingleValueDOMElement<Date, DatePicker> {
 
-    private static final int HEIGHT = 16;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/datepicker/DatePickerDOMElement.java]: Remove "dead" code

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/listbox/MultiValueDOMElement.java [hunk 3]
```diff
@@ -16,82 +16,59 @@
...
 import org.uberfire.ext.wires.core.grids.client.widget.grid.GridWidget;
 import org.uberfire.ext.wires.core.grids.client.widget.layer.GridLayer;
 
 /**
  * A DOMElement that can contain multiple values; e.g. a multi-select ListBox.
  */
-public abstract class MultiValueDOMElement<T, W extends ListBox> extends BaseDOMElement<T, W> {
+public abstract class MultiValueDOMElement<T, W extends ListBox> extends BaseDTDOMElement<T, W> {
 
-    private static final int HEIGHT = 16;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/listbox/MultiValueDOMElement.java]: Remove "dead" code and decide whether to limit the size of the editor to the cell size or let it _overflow_ (when "multi-select" ```ListBox``` at the moment).

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/textbox/SingleValueDOMElement.java [hunk 4]
```diff
@@ -17,81 +17,49 @@
...
+              gridLayer,
+              gridWidget);
+        style(widget).setWidth(100,
+                               Style.Unit.PCT);
+        style(widget).setHeight(100,
+                                Style.Unit.PCT);
+        style(widgetContainer).setPadding(2,
+                                          Style.Unit.PX);
 
-    public SingleValueDOMElement( final W widget,
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/textbox/SingleValueDOMElement.java]: Remove "dead" code and limit the editor to the size of the cell.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java [hunk 5]
```diff
@@ -257,50 +256,15 @@ public ListBox createWidget() {
                                          @Override
                                          public void toWidget(final GridCell<String> cell,
                                                               final ListBox widget) {
-                                             if (cell == null || cell.getValue() == null || cell.getValue().getValue() == null) {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java]: Delegate to a _helper_ to maximise re-use and better support testing.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java [hunk 6]
```diff
@@ -257,50 +256,15 @@ public ListBox createWidget() {
...
-                                                 }
-                                             }
+                                             BaseColumnConverterUtilities.toWidget(isMultipleSelect,
+                                                                                   cell,
+                                                                                   widget);
                                          }
 
                                          @Override
                                          public String fromWidget(final ListBox widget) {
-                                             final StringBuilder sb = new StringBuilder();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java]: Delegate to a _helper_ to maximise re-use and better support testing.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java [hunk 7]
```diff
@@ -357,30 +321,15 @@ public String convert(final String value) {
                                                @Override
                                                public void toWidget(final GridCell<String> cell,
                                                                     final ListBox widget) {
-                                                   final String value = cell.getValue().getValue();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java]: Delegate to a _helper_ to maximise re-use and better support testing.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java [hunk 8]
```diff
@@ -357,30 +321,15 @@ public String convert(final String value) {
...
-                                                                              values.contains(widget.getValue(i)));
-                                                   }
+                                                   BaseColumnConverterUtilities.toWidget(true,
+                                                                                         cell,
+                                                                                         widget);
                                                }
 
                                                @Override
                                                public String fromWidget(final ListBox widget) {
-                                                   final StringBuilder sb = new StringBuilder();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterImpl.java]: Delegate to a _helper_ to maximise re-use and better support testing.

**Comment by manstis** [comment]
Part of an ensemble:

- https://github.com/AppFormer/uberfire/pull/838
- https://github.com/kiegroup/drools-wb/pull/624

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactory.java (hunk 1 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactory.java]: Looks reasonable

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    @SuppressWarnings("unchecked")
+    public void makeOnCreationCallbackListBoxMultipleSelect() {
+        setupMultipleSelectTest(CallbackFactory.MAX_VISIBLE_ROWS,
+                                true);
+
+        final Callback callback = CallbackFactory.makeOnCreationCallback(factory,
+                                                                         cell,
+                                                                         enumLookups);
+        callback.callback(multiValueDOMElement);
+
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java]: Any chance to add `verify(multiValueWidget).setVisibleItemCount(eq(CallbackFactory.MAX_VISIBLE_ROWS));`?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    @SuppressWarnings("unchecked")
+    public void makeOnCreationCallbackListBoxSingleSelect() {
+        setupMultipleSelectTest(CallbackFactory.MAX_VISIBLE_ROWS,
+                                false);
+
+        final Callback callback = CallbackFactory.makeOnCreationCallback(factory,
+                                                                         cell,
+                                                                         enumLookups);
+        callback.callback(multiValueDOMElement);
+
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java]: Any chance to add `verify(multiValueWidget, never()).setVisibleItemCount(anyInt());`?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterUtilities.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                                final GridCell<String> cell,
+                                final ListBox widget) {
+        if (cell == null || cell.getValue() == null || cell.getValue().getValue() == null) {
+            if (widget.getItemCount() > 0) {
+                widget.setSelectedIndex(0);
+            }
+        } else {
+            final String value = cell.getValue().getValue();
+            if (isMultipleSelect) {
+                final List<String> values = Arrays.asList(value.split(","));
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterUtilities.java]: Well, when I looked on this line I got and idea about corner case, when enumeration for string field contains ',' itself. Lets say, I would like to have table cell, that checks if name is contained in list: "Sir, Michael" or "Michael". I was not able to construct such two item list that would appear in the cell list box. Question is, should be I able?

Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterUtilities.java]: @jomarko There are numerous (well, at least one) JIRA about enums containing "," themselves being broken so whatever fix is made for that will touch this code in the future. For the purpose of this PR, use of "," in a value itself is broken and remains broken.

Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/BaseColumnConverterUtilities.java]: @manstis I agree, still my approval holds true.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java (hunk 10 above)
Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java]: Done.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java (hunk 9 above)
Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java]: Done.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java (hunk 10 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java]: Thanks

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java (hunk 9 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/CallbackFactoryTest.java]: Thanks

**Comment by manstis** [comment]
Jenkins please retest this.

**Pull request merged**